### PR TITLE
Use dir metadata for top hash

### PR DESCRIPTION
**Pull request by sir-sigurd**
## TODO

- [x] Unit Tests

### CODE REVIEW

File: api/python/quilt3/packages.py [hunk 1]
```diff
@@ -1139,20 +1139,28 @@ def top_hash(self):
         """
```
Comment (akarve) [inline api/python/quilt3/packages.py:1139]: Please add the high-level top_hash algorithm to the docs here (i.e. note that metadata are hashed as of <commit>).

File: api/python/quilt3/packages.py [hunk 2]
```diff
@@ -1139,20 +1139,28 @@ def top_hash(self):
...
                     "PackageEntry missing hash and/or size: %s" % entry.physical_key
                 )
-            entry_dict = entry.as_dict()
-            entry_dict['logical_key'] = logical_key
-            entry_dict.pop('physical_keys', None)
-            entry_dict_str = json.dumps(entry_dict, sort_keys=True, separators=(',', ':'))
-            top_hash.update(entry_dict_str.encode('utf-8'))
-
-        return top_hash.hexdigest()
+            yield {
```
Comment (akarve) [inline api/python/quilt3/packages.py:1158]: Much cleaner.

File: api/python/tests/integration/test_packages.py [hunk 3]
```diff
@@ -863,6 +863,15 @@ def test_top_hash_stable(self):
         assert pkg.top_hash == top_hash, \
             "Unexpected top_hash for {}/packages/.quilt/packages/{}".format(registry, top_hash)
 
+    def test_top_hash_stable_dir_meta(self):
+        registry = DATA_DIR.as_posix()
+        top_hash = '20de5433549a4db332a11d8d64b934a82bdea8f144b4aecd901e7d4134f8e733'
```
Comment (akarve) [inline api/python/tests/integration/test_packages.py:868]: this is fine but the unit test will break every time someone modifies `/data`. So I'd use set_dir here, get the `.top_hash` and then assert that the hash changes after calling `.set_meta`.

then maybe patch `get_top_hash_parts` (which doesn't have any test coverage) and return whatever you want so you can check against a stable hash?

File: api/python/quilt3/packages.py (hunk 1 above)
Comment (sir-sigurd) [reply api/python/quilt3/packages.py:1139]: > note that _get_top_hash_parts indicates what from each entry contributes to the hash

This docstring is public, why mention a private method here?

Comment (akarve) [reply api/python/quilt3/packages.py:1139]: Agreed to skip mention it on the docstring

File: api/python/tests/integration/test_packages.py (hunk 3 above)
Comment (sir-sigurd) [reply api/python/tests/integration/test_packages.py]: I merged the new test with `test_top_hash_stable()`. We could completely inline the [test manifest](https://github.com/quiltdata/quilt/blob/master/api/python/tests/integration/data/.quilt/packages/20de5433549a4db332a11d8d64b934a82bdea8f144b4aecd901e7d4134f8e733) in the test if it makes sense to you.

**Comment by akarve** [comment]
Per discussion we should merge this as part for quilt3 4.x so as not to merge a potentially breaking change into 3.x.

**Comment by drernie** [comment]
@sir-sigurd This seems outdated and of unclear value.  Closing, unless you want to revive and merge.

**Pull request closed without merging**